pyteal/compiler/sourcemap.py

```diff
@@ -0,0 +1,985 @@
+            _TEAL_LINE_NUMBER: self.teal_lineno,
+            _TEAL_LINE: self.teal_line,
+            _PROGRAM_COUNTERS: self.pcs_repr(),
+            _PYTEAL_HYBRID_UNPARSED: self.hybrid_unparsed(),
+            _PYTEAL_NODE_AST_UNPARSED: self.node_source(),
+            _PYTEAL_NODE_AST_QUALNAME: self.code_qualname(),
+            _PYTEAL_COMPONENT: self.teal_component,
+            _PYTEAL_NODE_AST_SOURCE_BOUNDARIES: self.node_source_window(),
+            _PYTEAL_FILENAME: self.file(),
+            _PYTEAL_LINE_NUMBER: self.lineno(),
+            _PYTEAL_LINE_NUMBER_END: self.node_end_lineno(),
+            _PYTEAL_COLUMN: self.column(),
+            _PYTEAL_COLUMN_END: self.node_end_col_offset(),
+            _PYTEAL_LINE: self.code(),
+            _PYTEAL_NODE_AST: self.node,
+            _PYTEAL_NODE_AST_NONE: self.failed_ast(),
+            _STATUS_CODE: self.status_code(),
+            _STATUS: self.status(),
+        }
+
+        assert (
+            kwargs.keys() <= attrs.keys()
+        ), f"unrecognized parameters {kwargs.keys() - attrs.keys()}"
+
+        return OrderedDict(((kwargs[k], attrs[k]) for k in kwargs))
+
+    def validate_for_export(self) -> None:
+        """
+        Ensure providing necessary and unambiguous data before exporting.
+        """
+        if self.teal_lineno is None:
+            raise ValueError("unable to export without valid line number: None")
+        if (line := self.lineno()) is None or self.column() is None:
+            col = self.column()
+            raise ValueError(
+                f"unable to export without valid line and column for SOURCE but got: {line=}, {col=}"
+            )
+
+    def source_mapping(self, _hybrid: bool = True) -> "R3SourceMapping":
+        self.validate_for_export()
+        return R3SourceMapping(
+            line=cast(int, self.teal_lineno) - 1,
+            column=0,
+            column_end=len(self.teal_line) - 1,
+            source=self.file(),
+            source_line=cast(int, self.lineno()) - 1,
+            source_column=self.column(),
+            source_line_end=nel - 1 if (nel := self.node_end_lineno()) else None,
+            source_column_end=self.node_end_col_offset(),
+            source_extract=self.hybrid_unparsed() if _hybrid else self.code(),
+            target_extract=self.teal_line,
+        )
+
+
+@dataclass(frozen=True)
+class PyTealSourceMap:
+    r3_sourcemap: R3SourceMap | None
+    pc_sourcemap: PCSourceMap | None
+    annotated_teal: str | None
+
+
+class _PyTealSourceMapper:
+    """The workhorse class that runs the sourcemapping algorithm.
+    User-safe Expr-less artifacts are provided by methods:
+    * get_sourcemap()
+    * as_r3sourcemap()
+    * annotated_teal()
+    """
+
+    def __init__(
+        self,
+        teal_chunks: list[str],
+        components: list["pt.TealComponent"],
+        *,
+        teal_file: str | None = None,
+        include_pcs: bool = False,
+        algod: AlgodClient | None = None,
```

Comment:
I think introducing an AlgodClient dependency in PyTeal sounds a bit tricky:
* It increases testing and maintenance burden - we would have more tests that need to rely on algod, which is slower to write and run
* It feels unintuitive and perhaps un-modular to require a node for a compiler to generate extra debug symbols (and unfortunately that is still the case for compiling TEAL code, and maybe this can be excused as well in this context)

I do like the idea of including this feature in a separate SDK or tool if possible, so some set of debug symbols can be used by an external tool to call the SDK/algod and generate the PC mapping as well. Also open to other thoughts as well.

Reply:
You bring up good points. I'll address each in turn:

> It increases testing and maintenance burden - we would have more tests that need to rely on algod, which is slower to write and run
In this repo, I'm introducing 2 new integration tests:
* `tests/integration/algod_test.py`
* `tests/integration/sourcemap_monkey_integ_test.py`

The first is a test mostly of `algod_with_assertion`. A version of this function actually pre-existed this PR, but as it's being exposed in a non-testing code path, I felt it was important to introduce such a test. However, **algod** is already an integral part of the integration tests which employ dry-runs through "graviton".

The second is test of the pc-source-mapping functionality.

Compared to other pre-existing integration tests in the repo, they are lightweight.

So in my opinion, the increased _maintenance burden_  for introducing `AlgodClient` officially into PyTeal is marginal and worthwhile.

> It feels unintuitive and perhaps un-modular to require a node for a compiler to generate extra debug symbols (and unfortunately that is still the case for compiling TEAL code, and maybe this can be excused as well in this context)
I agree with this sentiment. But this the currently an easy way to produce PC's. If a better way arises in the future, we should swap it out.